POEM_051.md

```diff
@@ -0,0 +1,107 @@
+POEM ID:  051  
+Title:  Modifications to relative step sizing in finite difference  
+authors: [kenneth-t-moore]  
+Competing POEMs: N/A  
+Related POEMs: N/A  
+Associated implementation PR:  
+
+Status:
+
+- [x] Active
+- [ ] Requesting decision
+- [ ] Accepted
+- [ ] Rejected
+- [ ] Integrated
+
+
+
+Motivation
+----------
+
+When a user selects 'rel' for the `step_calc` when performing a finite difference approximation, the
+stepsize for a scalar variable is computed by taking the given `step` and multiplying it by the
+magnitude of the variable's current value.  If that variable is a vector, the stepsize is multiplied
+by the `norm` of the entire vector. This can lead to unintuitive performance, because the norm
+value is a function of the vector size. This can be verified by computing the norm of a large vector
+of ones and comparing to the scalar value.  Taking the norm is not the proper way to compute a
+relative stepsize.  An alternative is needed that also addresses these issues:
+
+1. The requested step direction must be preserved independent of the sign of the local value.
+
+2. The computed step should never be zero.
+
+
+Description
+-----------
+
+This POEM proposes two new ways to compute the relative step.
+
+
+Use average absolute vector value instead of the norm.
+======================================================
+
+Compute the step as:
+
+```
+    rel_step = step * np.sum(np.abs(x)) / n
+```
+
+This fixes the bug that caused the relative step to be larger than expected for wide vectors. The
+absolute value ensures that we won't be taking a forward step when we request backward.
+
+
+Allow the user to specify a relative step for every element in the vector.
+==========================================================================
+
+This gives the user finer control over the approximation. For `declare_partials`, the simplest API
+would be to allow the user to pass a list (or tuple) with the same length as the input whose
+derivatives are being approximated. Here is a simple example for a 3-wide input 'x':
+
+```
+   self.declare_partials(of='*', wrt='x', method='fd',
+                         step=[1e-7, 1e-5, 1e-6], form='forward', step_calc='rel')
+```
+
+The key difference here is that relative step_size is computed individually for each vector element:
+```
+    rel_step_j = step_j * np.abs(x_j)
+```
+
+This API might be able to be expanded to allow specification of other aspects of the approximation:
```

Comment:
I don't think we need this level of flexibility. Just allowing for a more intuitive step size calc based on each element (instead of some aggregate of teh whole vector) is enough.